### DISCUSSION TITLE
4 moderate severity vulnerabilities. How to fix it?

### DISCUSSION THREAD

**Discussion by Juancafr** [npm/feedback#889] [Registry]
I'm very new to this topic and I'm trying to install an app from a friend. When I use the npm audit command it tells me that there are 4 problems so to speak, and when I try to install them separately the same thing happens, any ideas?

request  *
Severity: moderate
Server-Side Request Forgery in Request - https://github.com/advisories/GHSA-p8p7-x288-28g6
No fix available
node_modules/request
  node-telegram-bot-api  *
  Depends on vulnerable versions of request
  Depends on vulnerable versions of request-promise
  node_modules/node-telegram-bot-api
  request-promise  >=0.0.2
  Depends on vulnerable versions of request
  Depends on vulnerable versions of request-promise-core
  node_modules/request-promise
  request-promise-core  *
  Depends on vulnerable versions of request
  node_modules/request-promise-core

4 moderate severity vulnerabilities

Some issues need review, and may require choosing
a different dependency.

**Comment by ljharb** [comment]
There's nothing you *can* do except replace whatever direct dependencies transitively depend on `request`, or, ignore the warnings.

**Reply by Juancafr** [reply]
> There's nothing you _can_ do except replace whatever direct dependencies transitively depend on `request`, or, ignore the warnings.

And how are they replaced? I'm very new to this topic, sorry.

**Reply by ljharb** [reply]
`npm explain request` will tell you which top-level deps are bringing it in. Then, you can find alternatives.

**Reply by Juancafr** [reply]
> `npm explain request` will tell you which top-level deps are bringing it in. Then, you can find alternatives.

I don't understand anything


root@DietPi:/mnt/USB/VineMonitoringBot# npm explain request                     request@2.88.2
node_modules/request
  request@"^2.88.2" from the root project
  request@"^2.83.0" from node-telegram-bot-api@0.61.0
  node_modules/node-telegram-bot-api
    node-telegram-bot-api@"^0.61.0" from the root project
  peer request@"^2.34" from request-promise@4.2.6
  node_modules/request-promise
    request-promise@"^4.2.6" from the root project
    request-promise@"^4.2.2" from node-telegram-bot-api@0.61.0
    node_modules/node-telegram-bot-api
      node-telegram-bot-api@"^0.61.0" from the root project
  peer request@"^2.34" from request-promise-core@1.1.4
  node_modules/request-promise-core
    request-promise-core@"^1.1.4" from the root project
    request-promise-core@"1.1.4" from request-promise@4.2.6
    node_modules/request-promise
      request-promise@"^4.2.6" from the root project
      request-promise@"^4.2.2" from node-telegram-bot-api@0.61.0
      node_modules/node-telegram-bot-api
        node-telegram-bot-api@"^0.61.0" from the root project

**Reply by ljharb** [reply]
That says that the root project - your code - is using `request`, as well as `node-telegram-bot-api`, as well as `request-promise`, as well as `request-promise-core`, so you'd have a lot of refactoring to do to get rid of those.

**Reply by Juancafr** [reply]
> That says that the root project - your code - is using `request`, as well as `node-telegram-bot-api`, as well as `request-promise`, as well as `request-promise-core`, so you'd have a lot of refactoring to do to get rid of those.

And could you help me? I don't know almost anything about commands, I'm new

**Reply by ljharb** [reply]
This isn't the appropriate venue for that; I'd suggest Stack Overflow, or a Slack or Discord community, or asking whoever built the project in the first place.
